### ISSUE TITLE
Checksums outside of EasyConfigs

### ISSUE THREAD

**Issue by mboisson** [easybuilders/easybuild-framework#3746]
At the moment, checksums for files are specified in EasyConfig files. This means that for every single copy of a recipe made for different toolchains, or even different versions that use the same patches, we repeat this information. 

My proposal is to implement a mechanism to get checksums from outside of the EasyConfig file. 

The file would be named : 
`<software name>.checksums` (or `checksums.json`)
and it would contain a dictionary of this format: 
```
{
   'filename1': 'checksum1',
   'filename2': 'checksum2',
}
``` 
The file would be located inside the EasyConfig directory for that software, as so to not obscure where the information is coming from (global parameters can be obscure). 

It could include patches as well as checksums for various versions of a given software. 

This would allow to completely eliminate checksums from the EasyConfig files (although we could support having both, with the EasyConfig-specified checksums having the highest priority). 

This is a competing proposal to https://github.com/easybuilders/easybuild-framework/issues/3744

**Comment by ocaisa** [comment]
I like the general idea. It would also easily solve the issue we have of checksums changing over time (update a single location). It can also preserve backwards compatibility. 

One concern I would have is what to do with pull requests. If we had such a mechanism we would like to encourage/enforce it's use. That means that people would have to submit (at least) 2 files with every PR and there would be a high chance of a conflict in the sources yaml.

**Comment by mboisson** [comment]
Maybe I am naive, but the learning curve of doing a PR is much steeper than the learning curve of adding more than 1 file to a PR ?

It could also occur that there are PRs *only* on the checksum file. For example, there is no need to create a separate EasyConfig if the only difference is the software version, you just add the checksum of the new version's source file to the checksum file, and use `--try-software-version=...`

**Comment by ocaisa** [comment]
> Maybe I am naive, but the learning curve of doing a PR is much steeper than the learning curve of adding more than 1 file to a PR ?
> 

In general yes, but not with EB: `--new-pr`. Many, many people use that and this will require that they edit another file to make a PR. What we can do is make `--inject-checksums` do it for them but I don't see how you can get away from the likelihood of conflicts, and conflict resolution is not something we can handle right now.

> It could also occur that there are PRs _only_ on the checksum file. For example, there is no need to create a separate EasyConfig if the only difference is the software version, you just add the checksum of the new version's source file to the checksum file, and use `--try-software-version=...`
>
> Or in the case of evolving checksums, you just do a PR on the checksum file.

Sure, I get that, that's one of the reasons I like the idea, but this also highlights how the checksum file becomes a git conflict hotspot.

**Comment by mboisson** [comment]
`--new-pr` could automatically detect and include changed `checksums.yaml` files ?

**Comment by ocaisa** [comment]
What I mean with the conflicts issue is that if PRs for two versions of a package at the same time, they will both be targeting the end of `checksums.yaml`, and when one get's merged the other end's up with a conflict. Maybe a git guru can figure out a way to help avoid that, or `--inject-checksums` get's clever and can fix this issue for you.

**Comment by branfosj** [comment]
Keeping the checksums list in order (alphabetical seems sensible) would reduce the 'two people add to the end causing a git conflict'. I'd like it to be in some order just so I can read it.

A side benefit of this is that it'd stop inject checksums reordering the easyconfig.

**Comment by boegel** [comment]
I like this idea a lot too, especially that we can switch to it gradually, and that it doesn't have impact on people who don't adopt it (checksums specified in easyconfigs still win).

I don't consider this a competing proposal to #3744 btw: we can provide support for both. But for the central repository, a separate `checksums` file probably makes more sense, since it makes maintenance easier.

W.r.t. format: I wouldn't use YAML for this, since that's not supported by the Python standard library (and I really REALLY want to avoid required dependencies for the core functionality of EasyBuild).
JSON would work, and would probably be sufficient, that's natively supported in Python 2 + 3.

**Comment by mboisson** [comment]
I can switch the code to Json, but we already have `yaml` as a dependency, don't we ? 
```
[mboisson@build-node easybuild-framework]$ git grep "import yaml"
easybuild/framework/easyconfig/format/yeb.py:    import yaml
easybuild/framework/easystack.py:    import yaml
test/framework/yeb.py:    import yaml
test/framework/yeb.py:        from easybuild.framework.easyconfig.format.yeb import yaml_join  # noqa
```

**Comment by mboisson** [comment]
Regarding `--inject-checksums`, should it
1) keep injecting in EasyConfigs ?
2) switch to generating `checksums.json` ?
3) somehow give the option to do either ?

**Comment by branfosj** [comment]
> I can switch the code to Json, but we already have `yaml` as a dependency, don't we ?

In a clean virtualenv
```bash
$ pip install easybuild
$ python -c "import yaml"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'yaml'
$ pip freeze
easybuild==4.4.0
easybuild-easyblocks==4.4.0
easybuild-easyconfigs==4.4.0
easybuild-framework==4.4.0
```

It is an optional dependency that many of us developing easybuild will have but is not needed to use easybuild.

**Comment by mboisson** [comment]
I switched the PR to use json instead of yaml

**Comment by mboisson** [comment]
In PR https://github.com/easybuilders/easybuild-framework/pull/3749, I implemented `--inject-checksums-to-json`

**Comment by boegel** [comment]
closing this, support for using `checksums.json` has been merged via #3749
